src/dataframe/dataframe.jl

```diff
@@ -1090,14 +1090,61 @@ Base.convert(::Type{DataFrame}, A::AbstractMatrix) = DataFrame(A)
 
 Base.convert(::Type{DataFrame}, d::AbstractDict) = DataFrame(d)
 
+"""
+    push!(df::DataFrame, row::Union{AbstractDict, NamedTuple})
 
-##############################################################################
-##
-## push! a row onto a DataFrame
-##
-##############################################################################
+Add one row at the end of `df` in-place, using values from `row`.
+
+Values in `row` are matched to columns in `df` based on names (order is ignored).
+`row` may contain more columns than `df`, but all column names that are present
+in `df` must be present in `row`.
+
+Column types of `df` are preserved, and new values are converted if necessary.
+An error is thrown if conversion fails.
+
+As a special case, if `df` has no columns and `row` is a `NamedTuple`,
+columns are created for all values in `row`, using their names and order.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(A=1:3, B=1:3)
+3×2 DataFrame
+│ Row │ A     │ B     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 1     │
+│ 2   │ 2     │ 2     │
+│ 3   │ 3     │ 3     │
 
+julia> push!(df, (C="something", A=true, B=false))
+4×2 DataFrame
+│ Row │ A     │ B     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 1     │
+│ 2   │ 2     │ 2     │
+│ 3   │ 3     │ 3     │
+│ 4   │ 1     │ 0     │
+
+julia> push!(df, Dict(:A=>1.0, :B=>2.0, :C=>"something"))
+5×2 DataFrame
+│ Row │ A     │ B     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 1     │
+│ 2   │ 2     │ 2     │
+│ 3   │ 3     │ 3     │
+│ 4   │ 1     │ 0     │
+│ 5   │ 1     │ 2     │
+```
+"""
 function Base.push!(df::DataFrame, row::Union{AbstractDict, NamedTuple})
+    if ncol(df) == 0 && row isa NamedTuple
+        for (n, v) in pairs(row)
+            setproperty!(df, n, [v])
```

Comment:
Probably better use `Tables.allocatecolumn`, which matters if you get a `CategoricalValue`.

Reply:
Fixed (but please check it 😄)